File: services/celery-gateway/src/celery_gateway/routers/beats.py
from __future__ import annotations

import json
import logging
import uuid
from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import JSONResponse
from sqlalchemy import delete, desc, select, update

from ..db import get_session
from ..db.models import BeatRun, BeatSchedule
from ..middleware.auth import require_auth
from ..models.beats import (
    BeatRunResponse,
    BeatScheduleResponse,
    CreateBeatInput,
    UpdateBeatInput,
)
from ..services.beat_scheduler import (
    compute_next_run_at,
    dispatch_task,
    validate_cron_expression,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/beats", tags=["beats"])


@router.get("", response_model=list[BeatScheduleResponse])
async def list_beats() -> Any:
    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).order_by(desc(BeatSchedule.created_at))
        )
        return result.scalars().all()


@router.post("", response_model=None, dependencies=[Depends(require_auth)])
async def create_beat(body: CreateBeatInput) -> JSONResponse:
    args_raw = body.args or "[]"
    kwargs_raw = body.kwargs or "{}"
    queue = body.queue or "celery"
    enabled = body.enabled if body.enabled is not None else True

    # Cross-field validation: schedule-type-specific requirements
    if body.schedule_type == "interval":
        if not body.interval_seconds or body.interval_seconds <= 0:
            raise HTTPException(
                status_code=400,
                detail="Interval seconds must be a positive number",
            )
    elif body.schedule_type == "cron":
        if not body.cron_expression:
            raise HTTPException(
                status_code=400,
                detail="Cron expression is required",
            )
        cron_error = validate_cron_expression(body.cron_expression)
        if cron_error:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid cron: {cron_error}",
            )

    beat_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc)

    next_run_at: datetime | None = None
    if enabled:
        try:
            next_run_at = compute_next_run_at(
                body.schedule_type, body.interval_seconds, body.cron_expression
            )
        except (ValueError, TypeError) as exc:
            raise HTTPException(
                status_code=400,
                detail=f"Failed to compute next run: {exc}",
            ) from exc

    async with get_session() as session:
        beat = BeatSchedule(
            id=beat_id,
            name=body.name,
            task_names=json.dumps(body.task_names),
            args=args_raw,
            kwargs=kwargs_raw,
            queue=queue,
            schedule_type=body.schedule_type,
            interval_seconds=body.interval_seconds,
            cron_expression=body.cron_expression,
            enabled=enabled,
            max_run_count=body.max_run_count,
            total_run_count=0,
            next_run_at=next_run_at,
            created_at=now,
            updated_at=now,
        )
        session.add(beat)
        await session.commit()

    return JSONResponse({"id": beat_id}, status_code=201)


@router.get("/{beat_id}", response_model=BeatScheduleResponse)
async def get_beat(beat_id: str) -> Any:
    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).where(BeatSchedule.id == beat_id).limit(1)
        )
        beat = result.scalar_one_or_none()
        if not beat:
            raise HTTPException(status_code=404, detail="Beat not found")
        return beat


@router.put("/{beat_id}", response_model=None, dependencies=[Depends(require_auth)])
async def update_beat(beat_id: str, body: UpdateBeatInput) -> JSONResponse:
    updates = body.model_dump(exclude_unset=True)

    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).where(BeatSchedule.id == beat_id).limit(1)
        )
        existing = result.scalar_one_or_none()
        if not existing:
            raise HTTPException(status_code=404, detail="Beat not found")

        task_names = updates.get("task_names")
        if task_names is not None and len(task_names) == 0:
            raise HTTPException(
                status_code=400,
                detail="At least one task must be selected",
            )

        schedule_type = updates.get("schedule_type", existing.schedule_type)
        interval_seconds = updates.get("interval_seconds", existing.interval_seconds)
        cron_expression = updates.get("cron_expression", existing.cron_expression)

        if schedule_type == "cron" and cron_expression:
            cron_error = validate_cron_expression(cron_expression)
            if cron_error:
                raise HTTPException(
                    status_code=400,
                    detail=f"Invalid cron: {cron_error}",
                )

        enabled = updates.get("enabled", existing.enabled)
        now = datetime.now(timezone.utc)

        next_run_at = existing.next_run_at
        if any(
            k in updates
            for k in ("schedule_type", "interval_seconds", "cron_expression")
        ):
            if enabled:
                try:
                    next_run_at = compute_next_run_at(
                        schedule_type, interval_seconds, cron_expression
                    )
                except (ValueError, TypeError):
                    next_run_at = None

        values: dict[str, Any] = {
            "name": updates.get("name", existing.name),
            "task_names": (
                json.dumps(task_names) if task_names is not None else existing.task_names
            ),
            "args": updates.get("args", existing.args),
            "kwargs": updates.get("kwargs", existing.kwargs),
            "queue": updates.get("queue", existing.queue),
            "schedule_type": schedule_type,
            "interval_seconds": interval_seconds,
            "cron_expression": cron_expression,
            "enabled": enabled,
            "max_run_count": updates.get("max_run_count", existing.max_run_count),
            "next_run_at": next_run_at,
            "updated_at": now,
        }

        await session.execute(
            update(BeatSchedule).where(BeatSchedule.id == beat_id).values(**values)
        )
        await session.commit()

    return JSONResponse({"ok": True})


@router.delete("/{beat_id}", response_model=None, dependencies=[Depends(require_auth)])
async def delete_beat(beat_id: str) -> JSONResponse:
    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).where(BeatSchedule.id == beat_id).limit(1)
        )
        if not result.scalar_one_or_none():
            raise HTTPException(status_code=404, detail="Beat not found")
        await session.execute(
            delete(BeatSchedule).where(BeatSchedule.id == beat_id)
        )
        await session.commit()
    return JSONResponse({"ok": True})


@router.post("/{beat_id}/toggle", response_model=None, dependencies=[Depends(require_auth)])
async def toggle_beat(beat_id: str) -> JSONResponse:
    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).where(BeatSchedule.id == beat_id).limit(1)
        )
        existing = result.scalar_one_or_none()
        if not existing:
            raise HTTPException(status_code=404, detail="Beat not found")

        new_enabled = not existing.enabled
        now = datetime.now(timezone.utc)

        next_run_at = existing.next_run_at
        if new_enabled and not next_run_at:
            try:
                next_run_at = compute_next_run_at(
                    existing.schedule_type,
                    existing.interval_seconds,
                    existing.cron_expression,
                )
            except (ValueError, TypeError):
                pass

        await session.execute(
            update(BeatSchedule)
            .where(BeatSchedule.id == beat_id)
            .values(
                enabled=new_enabled,
                next_run_at=next_run_at if new_enabled else None,
                updated_at=now,
            )
        )
        await session.commit()

    return JSONResponse({"enabled": new_enabled})


@router.post("/{beat_id}/run-now", response_model=None, dependencies=[Depends(require_auth)])
async def run_beat_now(beat_id: str) -> JSONResponse:
    async with get_session() as session:
        result = await session.execute(
            select(BeatSchedule).where(BeatSchedule.id == beat_id).limit(1)
        )
        beat = result.scalar_one_or_none()
        if not beat:
            raise HTTPException(status_code=404, detail="Beat not found")

        task_names: list[str] = json.loads(beat.task_names or "[]")
        if not task_names:
            raise HTTPException(status_code=400, detail="No tasks configured")

        args: list[Any] = json.loads(beat.args or "[]")
        kwargs: dict[str, Any] = json.loads(beat.kwargs or "{}")
        queue = beat.queue or "celery"
        now = datetime.now(timezone.utc)
        dispatched: list[str] = []

        for task_name in task_names:
            task_id: str | None = None
            error: str | None = None
            status = "SENT"

            try:
                task_id = await dispatch_task(task_name, args, kwargs, queue)
                dispatched.append(task_id)
            except Exception as exc:
                error = str(exc)
                status = "FAILURE"

            run = BeatRun(
                id=str(uuid.uuid4()),
                schedule_id=beat.id,
                task_id=task_id,
                task_name=task_name,
                args=beat.args,
                kwargs=beat.kwargs,
                queue=beat.queue,
                status=status,
                error=error,
                scheduled_at=now,
                sent_at=now,
            )
            session.add(run)

        await session.execute(
            update(BeatSchedule)
            .where(BeatSchedule.id == beat.id)
            .values(
                last_run_at=now,
                total_run_count=(beat.total_run_count or 0) + 1,
                updated_at=now,
            )
        )
        await session.commit()

    return JSONResponse({"taskIds": dispatched})


@router.get("/{beat_id}/runs", response_model=list[BeatRunResponse])
async def get_beat_runs(
    beat_id: str,
    limit: int = Query(50, ge=1, le=200),
) -> Any:
    async with get_session() as session:
        result = await session.execute(
            select(BeatRun)
            .where(BeatRun.schedule_id == beat_id)
            .order_by(desc(BeatRun.sent_at))
            .limit(limit)
        )
        return result.scalars().all()

File: packages/server/src/routes/beats.ts
import { Hono } from "hono";
import { getDb } from "../lib/db/index.js";
import { beatSchedules, beatRuns } from "../lib/db/schema.js";
import { computeNextRunAt, validateCronExpression } from "../lib/scheduler/cron.js";
import { eq, desc } from "drizzle-orm";
import { v4 as uuidv4 } from "uuid";
import { gatewaySendTask } from "../lib/celery-gateway.js";
import { sendCeleryTask } from "../lib/celery.js";
import { startScheduler, isSchedulerRunning } from "../lib/scheduler/index.js";

const app = new Hono();

interface CreateBeatInput {
  name: string;
  taskNames: string[];
  args?: string;
  kwargs?: string;
  queue?: string;
  scheduleType: "interval" | "cron";
  intervalSeconds?: number;
  cronExpression?: string;
  enabled?: boolean;
  maxRunCount?: number | null;
}

async function dispatchOneTask(
  taskName: string,
  args: unknown[],
  kwargs: Record<string, unknown>,
  queue: string
): Promise<string> {
  try {
    const res = await gatewaySendTask({
      task_name: taskName,
      args,
      kwargs,
      queue,
    });
    return res.task_id;
  } catch {
    return await sendCeleryTask(taskName, args, kwargs, queue);
  }
}

// GET /api/beats — list all beats
app.get("/", async (c) => {
  if (!isSchedulerRunning()) startScheduler();
  const db = getDb();
  const beats = db
    .select()
    .from(beatSchedules)
    .orderBy(desc(beatSchedules.createdAt))
    .all();
  return c.json(beats);
});

// POST /api/beats — create beat
app.post("/", async (c) => {
  const input = await c.req.json<CreateBeatInput>().catch(() => null);
  if (!input) {
    return c.json({ error: "Invalid JSON body" }, 400);
  }

  if (!input.name) {
    return c.json({ error: "Name is required" }, 400);
  }
  if (!input.taskNames || input.taskNames.length === 0) {
    return c.json({ error: "At least one task must be selected" }, 400);
  }

  if (input.scheduleType === "interval") {
    if (!input.intervalSeconds || input.intervalSeconds <= 0) {
      return c.json({ error: "Interval seconds must be a positive number" }, 400);
    }
  } else if (input.scheduleType === "cron") {
    if (!input.cronExpression) {
      return c.json({ error: "Cron expression is required" }, 400);
    }
    const cronError = validateCronExpression(input.cronExpression);
    if (cronError) {
      return c.json({ error: `Invalid cron: ${cronError}` }, 400);
    }
  } else {
    return c.json({ error: "Schedule type must be 'interval' or 'cron'" }, 400);
  }

  try {
    if (input.args) JSON.parse(input.args);
  } catch {
    return c.json({ error: "Invalid JSON for args" }, 400);
  }
  try {
    if (input.kwargs) JSON.parse(input.kwargs);
  } catch {
    return c.json({ error: "Invalid JSON for kwargs" }, 400);
  }

  const id = uuidv4();
  const now = new Date().toISOString();
  const enabled = input.enabled !== false;

  let nextRunAt: string | null = null;
  if (enabled) {
    try {
      nextRunAt = computeNextRunAt(
        input.scheduleType,
        input.intervalSeconds || null,
        input.cronExpression || null
      ).toISOString();
    } catch (e) {
      return c.json(
        { error: `Failed to compute next run: ${e instanceof Error ? e.message : e}` },
        400
      );
    }
  }

  if (!isSchedulerRunning()) startScheduler();
  const db = getDb();
  db.insert(beatSchedules)
    .values({
      id,
      name: input.name,
      taskNames: JSON.stringify(input.taskNames),
      args: input.args || "[]",
      kwargs: input.kwargs || "{}",
      queue: input.queue || "celery",
      scheduleType: input.scheduleType,
      intervalSeconds: input.intervalSeconds || null,
      cronExpression: input.cronExpression || null,
      enabled,
      maxRunCount: input.maxRunCount ?? null,
      totalRunCount: 0,
      nextRunAt,
      createdAt: now,
      updatedAt: now,
    })
    .run();

  return c.json({ id }, 201);
});

// GET /api/beats/:id — get single beat
app.get("/:id", async (c) => {
  const id = c.req.param("id");
  const db = getDb();
  const beat = db
    .select()
    .from(beatSchedules)
    .where(eq(beatSchedules.id, id))
    .get();

  if (!beat) {
    return c.json({ error: "Beat not found" }, 404);
  }
  return c.json(beat);
});

// PUT /api/beats/:id — update beat
app.put("/:id", async (c) => {
  const id = c.req.param("id");
  const input = await c.req.json<Partial<CreateBeatInput>>().catch(() => null);
  if (!input) {
    return c.json({ error: "Invalid JSON body" }, 400);
  }

  const db = getDb();
  const existing = db
    .select()
    .from(beatSchedules)
    .where(eq(beatSchedules.id, id))
    .get();

  if (!existing) return c.json({ error: "Beat not found" }, 404);

  if (input.taskNames !== undefined && input.taskNames.length === 0) {
    return c.json({ error: "At least one task must be selected" }, 400);
  }

  const scheduleType = input.scheduleType || existing.scheduleType;
  const intervalSeconds =
    input.intervalSeconds !== undefined
      ? input.intervalSeconds
      : existing.intervalSeconds;
  const cronExpression =
    input.cronExpression !== undefined
      ? input.cronExpression
      : existing.cronExpression;

  if (scheduleType === "cron" && cronExpression) {
    const cronError = validateCronExpression(cronExpression);
    if (cronError) return c.json({ error: `Invalid cron: ${cronError}` }, 400);
  }

  if (input.args) {
    try {
      JSON.parse(input.args);
    } catch {
      return c.json({ error: "Invalid JSON for args" }, 400);
    }
  }
  if (input.kwargs) {
    try {
      JSON.parse(input.kwargs);
    } catch {
      return c.json({ error: "Invalid JSON for kwargs" }, 400);
    }
  }

  const enabled = input.enabled !== undefined ? input.enabled : existing.enabled;
  const now = new Date().toISOString();

  let nextRunAt = existing.nextRunAt;
  if (
    input.scheduleType !== undefined ||
    input.intervalSeconds !== undefined ||
    input.cronExpression !== undefined
  ) {
    if (enabled) {
      try {
        nextRunAt = computeNextRunAt(
          scheduleType as "interval" | "cron",
          intervalSeconds || null,
          cronExpression || null
        ).toISOString();
      } catch {
        nextRunAt = null;
      }
    }
  }

  db.update(beatSchedules)
    .set({
      name: input.name || existing.name,
      taskNames:
        input.taskNames !== undefined
          ? JSON.stringify(input.taskNames)
          : existing.taskNames,
      args: input.args !== undefined ? input.args : existing.args,
      kwargs: input.kwargs !== undefined ? input.kwargs : existing.kwargs,
      queue: input.queue !== undefined ? input.queue : existing.queue,
      scheduleType: scheduleType,
      intervalSeconds: intervalSeconds || null,
      cronExpression: cronExpression || null,
      enabled,
      maxRunCount:
        input.maxRunCount !== undefined ? input.maxRunCount : existing.maxRunCount,
      nextRunAt,
      updatedAt: now,
    })
    .where(eq(beatSchedules.id, id))
    .run();

  return c.json({ ok: true });
});

// DELETE /api/beats/:id — delete beat
app.delete("/:id", async (c) => {
  const id = c.req.param("id");
  const db = getDb();
  db.delete(beatSchedules).where(eq(beatSchedules.id, id)).run();
  return c.json({ ok: true });
});

// POST /api/beats/:id/toggle — toggle beat
app.post("/:id/toggle", async (c) => {
  const id = c.req.param("id");
  const db = getDb();
  const existing = db
    .select()
    .from(beatSchedules)
    .where(eq(beatSchedules.id, id))
    .get();

  if (!existing) return c.json({ error: "Beat not found" }, 404);

  const newEnabled = !existing.enabled;
  const now = new Date().toISOString();

  let nextRunAt = existing.nextRunAt;
  if (newEnabled && !nextRunAt) {
    try {
      nextRunAt = computeNextRunAt(
        existing.scheduleType as "interval" | "cron",
        existing.intervalSeconds,
        existing.cronExpression
      ).toISOString();
    } catch {
      // leave null
    }
  }

  db.update(beatSchedules)
    .set({
      enabled: newEnabled,
      nextRunAt: newEnabled ? nextRunAt : null,
      updatedAt: now,
    })
    .where(eq(beatSchedules.id, id))
    .run();

  return c.json({ enabled: newEnabled });
});

// POST /api/beats/:id/run-now — run beat immediately
app.post("/:id/run-now", async (c) => {
  const id = c.req.param("id");
  const db = getDb();
  const beat = db
    .select()
    .from(beatSchedules)
    .where(eq(beatSchedules.id, id))
    .get();

  if (!beat) return c.json({ error: "Beat not found" }, 404);

  const taskNames: string[] = JSON.parse(beat.taskNames || "[]");
  if (taskNames.length === 0) return c.json({ error: "No tasks configured" }, 400);

  const args = JSON.parse(beat.args || "[]");
  const kwargs = JSON.parse(beat.kwargs || "{}");
  const queue = beat.queue || "celery";
  const now = new Date().toISOString();
  const dispatched: string[] = [];

  for (const taskName of taskNames) {
    let taskId: string | undefined;
    let error: string | undefined;
    let status = "SENT";

    try {
      taskId = await dispatchOneTask(taskName, args, kwargs, queue);
      dispatched.push(taskId);
    } catch (e) {
      error = e instanceof Error ? e.message : String(e);
      status = "FAILURE";
    }

    db.insert(beatRuns)
      .values({
        id: uuidv4(),
        scheduleId: beat.id,
        taskId: taskId || null,
        taskName,
        args: beat.args,
        kwargs: beat.kwargs,
        queue: beat.queue,
        status,
        error: error || null,
        scheduledAt: now,
        sentAt: now,
      })
      .run();
  }

  db.update(beatSchedules)
    .set({
      lastRunAt: now,
      totalRunCount: (beat.totalRunCount || 0) + 1,
      updatedAt: now,
    })
    .where(eq(beatSchedules.id, beat.id))
    .run();

  return c.json({ taskIds: dispatched });
});

// GET /api/beats/:id/runs — get beat run history
app.get("/:id/runs", async (c) => {
  const scheduleId = c.req.param("id");
  const limit = Math.min(
    Math.max(Number(c.req.query("limit")) || 50, 1),
    200,
  );
  const db = getDb();
  const runs = db
    .select()
    .from(beatRuns)
    .where(eq(beatRuns.scheduleId, scheduleId))
    .orderBy(desc(beatRuns.sentAt))
    .limit(limit)
    .all();
  return c.json(runs);
});

export default app;
